src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1162,47 +1155,127 @@ static ucs_status_t uct_ib_mlx5dv_check_dc(uct_ib_device_t *dev)
 
     srq_attr.attr.max_sge   = 1;
     srq_attr.attr.max_wr    = 1;
-    srq = ibv_create_srq(pd, &srq_attr);
-    if (srq == NULL) {
+    *srq_p = ibv_create_srq(*pd_p, &srq_attr);
+    if (*srq_p == NULL) {
         ucs_error("%s: ibv_create_srq() failed: %m", uct_ib_device_name(dev));
         status = UCS_ERR_IO_ERROR;
         goto out_destroy_cq;
     }
 
-    qp_attr.send_cq              = cq;
-    qp_attr.recv_cq              = cq;
-    qp_attr.qp_type              = IBV_QPT_DRIVER;
-    qp_attr.comp_mask            = IBV_QP_INIT_ATTR_PD;
-    qp_attr.pd                   = pd;
-    qp_attr.srq                  = srq;
+    return UCS_OK;
+
+out_destroy_cq:
+    ibv_destroy_cq(*cq_p);
+out_dealloc_pd:
+    ibv_dealloc_pd(*pd_p);
+out:
+    return status;
+}
 
-    dv_attr.comp_mask            = MLX5DV_QP_INIT_ATTR_MASK_DC;
-    dv_attr.dc_init_attr.dc_type = MLX5DV_DCTYPE_DCT;
-    dv_attr.dc_init_attr.dct_access_key = UCT_IB_KEY;
+static void uct_ib_mlx5dv_close_tmp_resources(struct ibv_srq *srq,
+                                              struct ibv_pd *pd,
+                                              struct ibv_cq *cq)
+{
+    uct_ib_destroy_srq(srq);
+    ibv_destroy_cq(cq);
+    ibv_dealloc_pd(pd);
+}
+
+static size_t
+uct_ib_mlx5dv_tx_wqe_ratio(struct ibv_qp *qp,
+                           const uct_ib_qp_init_attr_t *qp_init_attr,
+                           size_t *tx_wqe_ratio_p)
+{
+    uct_ib_mlx5dv_qp_t qp_info = {};
+    uct_ib_mlx5dv_t obj        = {};
+    ucs_status_t status;
+
+    obj.dv.qp.in  = qp;
+    obj.dv.qp.out = &qp_info.dv;
+
+    status = uct_ib_mlx5dv_init_obj(&obj, MLX5DV_OBJ_QP);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *tx_wqe_ratio_p = qp_info.dv.sq.wqe_cnt / qp_init_attr->cap.max_send_wr;
+    return UCS_OK;
+}
+
+static void uct_ib_mlx5dv_qp_init_attr(uct_ib_qp_init_attr_t *qp_init_attr,
+                                       struct ibv_srq *srq, struct ibv_pd *pd,
+                                       struct ibv_cq *cq,
+                                       enum ibv_qp_type qp_type)
+{
+    qp_init_attr->send_cq         = cq;
+    qp_init_attr->recv_cq         = cq;
+    qp_init_attr->srq             = srq;
+    qp_init_attr->qp_type         = qp_type;
+    qp_init_attr->sq_sig_all      = 0;
+#if HAVE_DECL_IBV_CREATE_QP_EX
+    qp_init_attr->comp_mask       = IBV_QP_INIT_ATTR_PD;
+    qp_init_attr->pd              = pd;
+#endif
+    qp_init_attr->cap.max_send_wr = 128;
+
+    if (qp_type == IBV_QPT_DRIVER) {
+        qp_init_attr->cap.max_recv_wr = 0;
+        uct_ib_mlx5_dci_qp_update_attr(qp_init_attr);
+    } else {
+        ucs_assertv(qp_type == IBV_QPT_UD, "qp_type %u", qp_type);
+        qp_init_attr->cap.max_recv_wr     = 128;
+        qp_init_attr->cap.max_send_sge    = 2;
+        qp_init_attr->cap.max_recv_sge    = 1;
+        qp_init_attr->cap.max_inline_data = 0;
+    }
```

Comment:
maybe remove from this func, as it is not really common thing and used only once in dc and ud?

Reply:
done